### ISSUE TITLE
Add `Data.Traversable.zipWithList`

### ISSUE THREAD

**Issue by ChickenProp** [haskell/core-libraries-committee#229]
```haskell
zipWithList :: Traversable t => (a -> b -> c) -> t a -> [b] -> Maybe (t c)
zipWithList f t xs = sequenceA $ snd $ mapAccumL go xs t
  where go []     _ = ([], Nothing)
        go (x:xs) e = (xs, Just $ f e x)
```

This has the behavior of running simultaneously through a `Traversable` and a list, and replacing each value in the `Traversable` with `f valFromTraversable valFromList`. If the list is too short, we return `Nothing`. If it's longer than necessary, extra elements are ignored.

Examples:

```haskell
zipWithList ((+) @Int) (Just 1) [1, 2, 3] -- Just (Just 2)
zipWithList ((+) @Int) (Just 1) [] -- Nothing

zipWithList (++) [ [], [True] ] [ [], [False] ] -- Just [ [], [True, False] ]
zipWithList (++) [ [], [True] ] [ [] ] -- Nothing
```

I'm not married to the name or argument order. I felt like if it was used inline, then in ``x `zipWithList` y`` you'd expect `x` to be the `Traversable` and `y` to be the list; but you can't use it inline like that, so shrug.

I dunno how often something like this is needed (though I needed it myself recently and was somewhat surprised it's not in base[^1]). Part of why I'm suggesting it is that I think it might help to understand `Traversable`. The documentation already explains that

[^1]: Uh, caveat that I subsequently removed it because I needed something more general, roughly of type `(a -> b -> c) -> t (Either x a) -> [b] -> Maybe (t (Either x c))` that would only take an element from the list when it saw a `Right`. Still, I'd have found it temporarily useful to have `zipWithList`, and looking up the implementation would have helped me write the more powerful thing.

> It is in fact possible to decompose a traversable structure `t a` into its shape (a.k.a. spine) of type `t ()` and its element list `[a]`. The original structure can be faithfully reconstructed from its spine and element list.

But it doesn't say how to do these operations. The decomposition is already easy enough, but the reconstruction is currently awkward. `zipWithList` makes it straightforward.

* Given `x :: t a`, the spine is `const () <$> x` (using the `Functor` instance; we could also use `() <$ x` or `() $> x` but I forget which is which)
* Given `x :: t a`, the element list is `toList x` (using the `Foldable` instance)
* Given `spine :: t ()` and `elementList :: [a]`, the original is `fromJust $ zipWithList (\() x -> x) spine elementList`.

I think the last example would be significantly harder to follow if `zipWithList` had to be inlined and written with just `mapAccumL`.

Another way of putting this is that

```haskell
zipWithList (const () <$> t) (toList t) == Just t
```

For pedagogy it might easier to to instead use some function `replaceElementList :: t a -> [b] -> Maybe (t b)`, where `replaceElementList = zipWithList (\_ x -> x)`. I wouldn't be opposed to also adding such a function. But I expect `zipWithList` to be more useful.

**Comment by tomjaguarpaw** [comment]
Something like this seems plausible, particularly if it's the most general version in a family (perhaps the version below) but I think I'd need to feel confident that there's a lot of code in the wild calling out for this before incorporating it into `base`. I'm not yet convinced of that.

```haskell
zipWithList ::
  Traversable t =>
  (a -> Either c (b -> c)) ->
  t a ->
  [b] ->
  Maybe (t c, [b])
```

**Comment by Bodigrim** [comment]
A typical application is "enumerate elements of a container", `t a -> t (Int, a)`. People mostly write it via converting `t a` to list and back, which is ugly and inefficient. 

My preference would be 
```
alignWithList :: Traversable t => (a -> Maybe b -> c) -> t a -> [b] -> t c
```
The name is inspired by [`alignWith`](https://hackage.haskell.org/package/semialign-1.3/docs/Data-Semialign.html#v:alignWith).

**Comment by ChickenProp** [comment]
Oh, I like `alignWithList`! Then `zipWithList f = sequenceA $ alignWithList (\a mb -> f a <$> mb)` (or `alignWithList (fmap . f)`), which could also be included or not. Untested but I guess the implementation would be

```haskell
alignWithList f t xs = snd $ mapAccumL go xs t
  where go []     e = ([], f e Nothing)
        go (x:xs) e = (xs, f e $ Just x)
```

> What does the `Maybe` do here?

If the list is shorter than the traversable, the trailing elements of the traversable would be paired with `Nothing`.

We could generalize this with `Either` too, maybe

```haskell
alignHolesWithList
  :: Traversable t
  => (a -> Either c (Maybe b -> c))
  -> t a
  -> [b]
  -> (t c, [b])
```

Still,

> I think I'd need to feel confident that there's a lot of code in the wild calling out for this before incorporating it into `base`. I'm not yet convinced of that.

Yeah, this is a good question to ask. I'm not sure I have much to answer it with (maybe I'll come up with something), but I'm interested to see if others do.

**Comment by mixphix** [comment]
@ChickenProp Have you considered `[semialign](https://hackage.haskell.org/package/semialign-1.3/docs/Data-Semialign.html)`?

**Comment by ChickenProp** [comment]
@mixphix I hadn't explicitly considered it, but it wouldn't work for the use case that inspired this proposal.

**Comment by Bodigrim** [comment]
@ChickenProp how would you like to proceed? Are you ready to distill a final proposal?

**Comment by blamario** [comment]
If we aim to _Parse, not Validate_, this is clearly calling for adding an `Infinite` list data type to the base library. Then we wouldn't need to worry about running out of list elements:

    zipWithStream :: Traversable t => (a -> b -> c) -> t a -> Infinite b -> t c

I'm not seriously advocating this right now, but I'd be interested in collecting the use cases.

**Comment by treeowl** [comment]
When you get `Nothing`, what are you going to do with it? To put a finer point on it, can anyone think of an application where getting `Nothing` isn't a programmer error that will land in `error`? That makes me a bit nervous. Would it be better to use an infinite stream or an explicit unfold? Either of those would handle enumeration just fine.

An aside: even if the `Traversable` is list-like, there won't be any laziness. The lazy version would give you a container full of `Maybe`s.

**Comment by Bodigrim** [comment]
FWIW I'm open to extend [`infinite-list`](https://hackage.haskell.org/package/infinite-list) with something like
```haskell
zipWithInfinite :: Data.Traversable.Traversable t => (a -> b -> c) -> Infinite a -> t b -> t c
zipWithInfinite f = (snd .) . Data.Traversable.mapAccumL (\(x :< xs) b -> (xs, f x b))
```

**Comment by ChickenProp** [comment]
> how would you like to proceed? Are you ready to distill a final proposal?

Honestly, not sure. There's various functions that I think would all be somewhat positive, but I'm not inclined to fight hard for any of them. It would be nice to get a sense of what's likely to be supported before picking one and asking for a vote.

If it looks like none of them would get approval, then I might just (at some point) add a few lines to the documentation and submit that for approval. (That needs a vote too, right?)

> When you get `Nothing`, what are you going to do with it? To put a finer point on it, can anyone think of an application where getting `Nothing` isn't a programmer error that will land in `error`? That makes me a bit nervous.

Admittedly, nothing comes to mind (which doesn't mean they don't exist). I don't particularly see a problem with it, though.

(I'd say that if no situation like this comes to mind, then the `zip` variants probably make more sense than the `align` variants. `align` gives better control over the list-too-short case, but if we don't think there's anything to do in that case, then we might as well go for the simpler `zip`.)

> Would it be better to use an infinite stream or an explicit unfold? Either of those would handle enumeration just fine.

Infinite streams work for enumeration, but are more limited in general since we can't always create one. (I guess we could use `cycle`, but if we're wrong about having enough items then we get a non-crashing bug (except a crash if we have zero elements)).

I'm not sure how we'd use an unfold here, can you elaborate?

Looking through https://hackage-search.serokell.io/?q=mapAccumL to see where things like this would be useful:

- [OpenGL](https://hackage-search.serokell.io/viewfile/OpenGL-3.0.3.0/src/Graphics/Rendering/OpenGL/GL/Tensor.hs#line-289) - I think this is an enumerate followed by an `fmap`, which could be done with `zipWithInfinite`.
- [OpenGLRaw](https://hackage-search.serokell.io/viewfile/OpenGLRaw-3.3.4.1/RegistryProcessor/src/MangledRegistry.hs#line-196) - hard to follow but I think either of the hole functions (with `a -> Either ...`) will work here. Uses an infinite list.
- [Tensor](https://hackage-search.serokell.io/viewfile/Tensor-1.1.0.2/src/Data/Tensor.hs#line-289) - looks the same as `OpenGL`.
- [TestExplode](https://hackage-search.serokell.io/viewfile/TestExplode-0.1.0.0/doc/examples/FinalIO.hs#line-59) - enumeration. Explicitly calls out `mapAccumL` as being complicated. (But it's just a list, so `zipWith` would have been fine...)
- [ad](https://hackage-search.serokell.io/viewfile/ad-4.5.4/src/Numeric/AD/Internal/Combinators.hs#line-28) - defines `zipWithT` basically the same as I originally defined `zipWithList` (but for `Foldable f => f b` instead of `[b]`). Haven't checked how often it uses it. Also uses it for enumeration many times, [e.g.](https://hackage-search.serokell.io/viewfile/ad-4.5.4/src/Numeric/AD/Internal/Forward/Double.hs#line-115)
- [alms](https://hackage-search.serokell.io/viewfile/alms-0.6.9/src/Util.hs#line-124) - `align[Holes]WithList` could be used for `mapHead` and `mapTail`, but I don't think it would actually be any clearer.
- [arith-encode](https://hackage-search.serokell.io/viewfile/arith-encode-1.0.2/src/Data/ArithEncode/Basic.hs#line-493) - enumeration (but again, just a list).
- [aws-kinesis-client](https://hackage-search.serokell.io/viewfile/aws-kinesis-client-0.4.0.2/src/Aws/Kinesis/Client/Producer.hs#line-308) - enumeration

...and I'm gonna stop there, at least for now.

So I'd say "enumerating a `Traversable`" is pretty common, if people don't want the other functions then something like

```haskell
enumerate :: Traversable t => t a -> t (Int, a)
enumerateWith :: Traversable t => (Int -> a -> b) -> t a -> t b
```

might be a worthy addition, though they don't satisfy my original purpose. The other functions here are more general than those, and there's a couple things in this list that could make use of them.

**Comment by Bodigrim** [comment]
> It would be nice to get a sense of what's likely to be supported before picking one and asking for a vote.

You can ask for non-binding opinions instead of a formal vote, but it is vital to keep the list of options short.

> If it looks like none of them would get approval, then I might just (at some point) add a few lines to the documentation and submit that for approval. (That needs a vote too, right?)

No, documentation changes do not need CLC approval, just submit an MR.

> I'm not sure how we'd use an unfold here, can you elaborate?

It would be `foo :: Traversable t => (a -> b -> c) -> a -> (a -> a) -> t b -> t c`, where `a` and `a -> a` specify the initial value and the unfolding function.

**Comment by treeowl** [comment]
> > I'm not sure how we'd use an unfold here, can you elaborate?
> 
> It would be `foo :: Traversable t => (a -> b -> c) -> a -> (a -> a) -> t b -> t c`, where `a` and `a -> a` specify the initial value and the unfolding function.

Close, but not quite. The seed can have a different type.

```haskell
foo :: Traversable t => (a -> b -> c) -> s -> (s -> (a, s)) -> t b -> t c
```

**Comment by treeowl** [comment]
The unfolding approach is likely to be more efficient in situations where there's a natural "state", because it can avoid creating conses and also thunks for the tails. I suspect the `Infinite` approach is likely to be more efficient, as well as much simpler, when elements are naturally produced in chunks (i.e., more than one element at a time), because it can make better use of the evaluation stack. Both approaches are good.

**Comment by ChickenProp** [comment]
> You can ask for non-binding opinions instead of a formal vote, but it is vital to keep the list of options short.

Noted. At least right now I'm happy letting people discuss.

On the unfolding functions: thanks. I gotta admit, I'm not wild about adding something that takes two function arguments to traverse a single structure :p I guess the thing these do compared to `mapAccumL` is forbid the state from updating based on the values in the container.

Oh, another search worth doing: stackage shows that `replaceElementList` is defined by [elynx-tree](https://hackage.haskell.org/package/elynx-tree-0.7.2.2) in two places. It's named [`setLeaves`](https://hackage.haskell.org/package/elynx-tree-0.7.2.2/docs/src/ELynx.Topology.Rooted.html#setLeaves) and [`setLabels`](https://hackage.haskell.org/package/elynx-tree-0.7.2.2/docs/src/ELynx.Tree.Rooted.html#setLabels), which both have the same general type despite the names sounding specialized.

For whatever it's worth, if we were to add one thing, I currently lean towards my original `zipWithList` as being my favorite. There's a tradeoff between "if it's too specific it's not very useful" and "if it's too general it's complicated to use and doesn't save much over `mapAccumL`. I like where `zipWithList` lands on that tradeoff. It's still easy to use for enumerations (you need `fromJust`, but if you know you have an infinite list that seems okay), but also has more general uses and seems pedagogically helpful.

**Comment by ChickenProp** [comment]
> It's still easy to use for enumerations (you need `fromJust`, but if you know you have an infinite list that seems okay)

Oh, damn, I spoke too soon. If you have an infinite list *and an infinite `Traversable`*, this will just never return :( it needs to fully traverse both structures to figure out if it returns `Just` or `Nothing`. `alignWithList` would be able to, or a variant that errors instead of returning `Maybe`, and `zipWithInfinite` would be fine. I think this makes me lean towards `alignWithList` instead of `zipWithList`.

I now realize treeowl said this above, I just didn't understand it at the time:

> An aside: even if the `Traversable` is list-like, there won't be any laziness. The lazy version would give you a container full of `Maybes`.

**Comment by ChickenProp** [comment]
Okay, no discussion for a bit so I'll go ahead and formally ask for a vote. I've edited the current proposal into the top post, but here it is as well. (Note that I've flipped the argument order from earlier discussion.)

```haskell
alignWithList :: Traversable t => (Maybe a -> b -> c) -> [a] -> t b -> t c
alignWithList f xs = snd . mapAccumL go xs
  where go []     y = ([], f Nothing y)
        go (x:xs) y = (xs, f (Just x) y)
```

This has the effect of running simultaneously through a `Traversable` and a list, and replacing each value in the `Traversable` with `f (Just valFromList) valFromTraversable`; or `f Nothing valFromTraversable` if the list has run out. If the list is longer than the `Traversable`, extra elements are ignored.

Examples:

```haskell
alignWithList (\mx y -> mx <&> (+ y)) [1, 2, 3] (Just 1) -- Just (Just 2)
alignWithList (\mx y -> mx <&> (+ y)) [] (Just 1) -- Just Nothing

alignWithList (\mx y -> mx <&> (++ y)) [ [], [False] ] [ [], [True] ] -- [ Just [], Just [False, True ] ]
alignWithList (\mx y -> mx <&> (++ y)) [ [] ] [ [], [True] ] -- [ Just [], Nothing ]
```

I don't propose to add these extra functions, but they can easily be defined in terms of it:

```haskell
zipWithListErr :: Traversable t => (a -> b -> c) -> [a] -> t b -> t c
zipWithListErr f = alignWithList (f . fromJust)

zipWithListMay :: Traversable t => (a -> b -> c) -> [a] -> t b -> Maybe (t c)
zipWithListMay f xs t =
  sequenceA $ alignWithList (\mx y -> f <$> mx <*> pure y) xs t

enumerate :: Traversable t => t a -> t (Int, a)
enumerate = alignWithList ((,) . fromJust) [0..]

enumerateWith :: Traversable t => (Int -> a -> b) -> t a -> t b
enumerateWith f = alignWithList (f . fromJust) [0..]

replaceElementListErr :: Traversable t => [a] -> t b -> t a
replaceElementListErr = alignWithList (const . fromJust)

replaceElementListMay :: Traversable t => [a] -> t b -> Maybe (t a)
replaceElementListMay xs t = sequenceA $ alignWithList const xs t
```

Of these, `enumerate` and `enumerateWith` are used fairly often in cabal (see [this comment](https://github.com/haskell/core-libraries-committee/issues/229#issuecomment-1890777502)). And I think `replaceElementList` has pedagogical value, helping to understand `Traversable`. The documentation already explains that

> It is in fact possible to decompose a traversable structure `t a` into its shape (a.k.a. spine) of type `t ()` and its element list `[a]`. The original structure can be faithfully reconstructed from its spine and element list.

But it doesn't say how to do these operations. The decomposition is already easy enough, but the reconstruction is currently awkward. `alignWithList` makes it straightforward.

* Given `x :: t a`, the spine is `const () <$> x` (using the `Functor` instance; we could also use `() <$ x` or `() $> x` but I forget which is which).
* Given `x :: t a`, the element list is `toList x` (using the `Foldable` instance).
* Given `spine :: t ()` and `elementList :: [a]`, the original is `alignWithList (\x _ -> fromJust x) elementList spine`.

I think the last example would be significantly harder to follow if `alignWithList` had to be inlined and written with just `mapAccumL`.

Another way of putting this is that

```
alignWithList (\x _ -> fromJust x) (const () <$> t) (toList t) == t
```

I currently like putting the list argument before the `Traversable` argument:

* Pro: I think in a chain like `x & f1 & ... & alignWithList f xs`, the natural reading is that "the thing being aligned with a list" is what gets passed around.
* Pro: It's easier to turn `f :: a -> b -> c` into `Maybe a -> b -> c` than into `a -> Maybe b -> c`. (These are `f . fromJust` and `\x -> f x . fromJust`.)
* Pro: This order more closely matches what I consider to be the natural type signatures of `enumerate[With]`, and the existing type signature of `mapAccum{L,R}`.
* Con: I think ``x `alignWithList` y`` would more naturally have the list being `y` (but you can't do this).
* Con: It's more awkward to turn `f :: a -> b -> c` into `Maybe a -> b -> Maybe c` than into `a -> Maybe b -> Maybe c`. (These are `\mx y -> f <$> mx <*> pure y` and `fmap . f`.)

**Comment by blamario** [comment]
`(Maybe a -> b -> c) ->` is isomorphic to but less efficient than `(a -> b -> c) -> (b -> c) ->`

**Comment by tomjaguarpaw** [comment]
-1

----

It hasn't been demonstrated that there's enough code in the wild calling out for this for me to want to include it in `base`.

**Comment by gbaz** [comment]
> "align" matches its use in the existing semialign package.

I don't think it does. There align "aligns" two structures of the same type. Further, `alignWith` like `zipWith` uses the `with` to describe the combining function used when two elements match up. Here, it appears that the `with` is used to refer more to the use of the list, rather than the function.

A clearer function, imho, would be `alignList :: Traversable t => [a] -> t b -> t (b, Maybe a)`. Then the `alignListWith` version would just be a composition with `fmap`.

That said, I'm not sure if I've had call for such a thing or if it belongs in base.

**Comment by ChickenProp** [comment]
Yeah, it's definitely using "with" differently, and "align" is maybe more "analogously with semialign" than "the same".

(I don't really want to bikeshed over the name. If it's a case of "people like the function but not the name and think maybe a minor variant would be better" then sure, we can try to figure out the best name and variant. If it's not getting in with any name or variant, there's not much point.)

**Comment by Bodigrim** [comment]
> Okay, no discussion for a bit so I'll go ahead and formally ask for a vote. I've edited the current proposal into the top post, but here it is as well. (Note that I've flipped the argument order from earlier discussion.)
> 
> ```haskell
> alignWithList :: Traversable t => (Maybe a -> b -> c) -> [a] -> t b -> t c
> alignWithList f xs = snd . mapAccumL go xs
>   where go []     y = ([], f Nothing y)
>         go (x:xs) y = (xs, f (Just x) y)
> ```

I'm afraid I'm lukewarm on this. I don't really like the `Maybe` bit: there does not seem to be many applications which would intentionally supply a finite list, shorter than the `Traversable` container. And yet it has to be present to avoid partiality. 

@ChickenProp how about we collaborate on https://github.com/Bodigrim/infinite-list/issues/15 first? If it works out and there is still a desire to get it into `base`, we can either move it along with `data Infinite a`, or replace `Infinite` with unfolding function. It would be possible to make a stronger case, because there will be a prior art and potentially a collection of real-world applications.

**Comment by ChickenProp** [comment]
(Just noting that I just wanted this again. I have a database function that, given a `[Foo]`, creates one `Bar` for each `Foo`; and I have a `Traversable t => t Foo`, and I wanted to turn that into a matching `t Bar`.)

> I don't really like the `Maybe` bit: there does not seem to be many applications which would intentionally supply a finite list, shorter than the `Traversable` container. And yet it has to be present to avoid partiality.

Agreed that this is annoying. I expect almost all the time, the user either passes in an infinite list/unfolding function (these seem pretty much equivalent to me, but I think `Infinite` is more ergonomic), or one at least as long as the `Traversable`. (Likely exactly as long.)

Thinking about it... If the type signature expects an `Infinite a`, and the user only has a `[a]` that they expect to be long enough, then they can always do something like ``x `Infinite.prependList` error "oops"``, or ``map Just x `Infinite.prependList` Infinite.repeat Nothing`` if they want to do actual error handling.

And if the type signature expects a `[a]`, and the user has an `Infinite a`, then `fromJust` is safe. If the user has a `[a]` that they expect to be infinite, or at least as long as the `Traversable`, then they can use either `fromJust` or do something else with the `Maybe` for some other form of error handling.

So they're equivalent in power, though I still think I prefer the "expect a `[a]`" side of the tradeoff.

> @ChickenProp how about we collaborate on https://github.com/Bodigrim/infinite-list/issues/15 first?

Seems reasonable. I mean, I'd still prefer to just get this in :p But that currently seems unlikely, so yeah.

**Comment by parsonsmatt** [comment]
I think the difficulty with the types and runtime errors here is a good indication that we may not be cutting reality at the joints here.

We're using the name `align`, which has [this `Semialign` class as metaphor](https://www.stackage.org/haddock/lts-22.11/semialign-1.3/Data-Align.html#t:Semialign). The function `alignWith :: f a -> f b -> f (These a b)` gives you a structure `f` that can be `This a`, `That b`, or `These a b` - allowing you to indicate *which* positions weren't matched.

So the question here seems to be: how do we take a `[a]` and make it into a `Semialign f => f a`, so that we can `align` it with another structure? [`Data.List.IsList`](https://www.stackage.org/haddock/lts-22.11/base/GHC-Exts.html#v:fromList) gives us a `fromList` candidate, but that may not be really what we want - `fromList` isn't necessarily total (ie `IsList (NonEmpty a)`). 

But on further thought, that's also not right. Consider an unbalanced binary tree datatype. 

```haskell
data BinTree a = Node a (Maybe (BinTree a)) (Maybe (BinTree a))

tree0 :: BinTree Int
tree0 = Node 0 (Just (Node 1 Nothing Nothing)) Nothing

tree1 :: BinTree Char
tree1 = Node 'a' Nothing (Just (Node 'b' Nothing Nothing))

>>> align tree0 tree1
Node (These 0 'a') (Just (Node (This 1) Nothing Nothing) (Just (Node (That 'b') Nothing Nothing))
```

The actual shape of the structure matters. The *method* we go from `[a]` to `BinTree a` matters. You can `leftLeaning, rightLeaning, balanced :: [a] -> BinTree a`, and none of them will generally work. So `align` doesn't really seem like a good intuition for this.

`zip` is close, but `zip :: [a] -> [b] -> [(a, b)]` is also not quite what we want - the type is tricky. For a total definition, we *truncate* the list. A more proper definition is probably `zip :: [a] -> [b] -> ( [(a, b)], Maybe (Either (NonEmpty a) (NonEmpty b)) )` - this gives us a second element `Nothing` if both lists are the same, or `Either` a nonempty containing the remainder elements. `zip :: [a] -> [b] -> [These a b]` is unsatisfying because we'll *definitely* have `These a b` for the list structure that is the same length, and we'll *definitely* have either `That a` or `This b` afterwards, so the type is not fully expressing what's going on.

---

There's *something* here, to be sure! But I don't think `alignWithList` is it. 

I think the proper next step for the concept is a library that defines it and tries to make a good abstraction around it. I do not think that it needs to be in `base`.

**Comment by Bodigrim** [comment]
@ChickenProp since you asked for a formal vote in https://github.com/haskell/core-libraries-committee/issues/229#issuecomment-1902115748, would you like to proceed with it or shall we mark the proposal as dormant for now?

**Comment by hasufell** [comment]
Interesting discussion, but I'd also vote -1 for now.

**Comment by parsonsmatt** [comment]
@mixphix  I must admit, that was a good nersnipe - while we can't directly package classes and patterns, we *can* make pattern synonyms that are polymoprhic in a class constraint:

```haskell
{-# language FunctionalDependencies, MultiParamTypeClasses, PatternSynonyms, ViewPatterns #-}

class ConsC x xs | xs -> x where
    cons :: x -> xs -> xs
    uncons :: xs -> Maybe (x, xs)

pattern ConsP :: ConsC x xs => () => x -> xs -> xs
pattern ConsP x xs <- (uncons -> Just (x, xs) ) where
    ConsP x xs = cons x xs
```

The signature on `uncons` should probably be `Maybe (x, Maybe xs)`, because `uncons :: NonEmpty a -> (a, [a])` implies that you can't always retrieve an element. But that's already enough of a digression.

**Comment by ChickenProp** [comment]
As a followup: I've now published a package [hetero-zip](https://hackage.haskell.org/package/hetero-zip-0.1.0.0) which offers this. (Thanks @Bodigrim for the name, and @parsonsmatt for making me realize I could publish as a separate package - for some reason that hadn't occured to me until your comment.)

**Comment by Bodigrim** [comment]
> A `Traversable` structure `x :: t a` can be decomposed into
> 
>   * its spine: `const () <$> x :: t ()`
>   * its element list: `toList x :: [a]`
>
> It can then be reconstructed from these. 

That's very insightful, thanks!
